src/data_watcher.hpp

```diff
@@ -0,0 +1,173 @@
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
+
+#include <sys/inotify.h>
+#include <unistd.h>
+
+#include <sdbusplus/async.hpp>
+
+#include <filesystem>
+
+namespace data_sync
+{
+namespace watch::inotify
+{
+
+namespace fs = std::filesystem;
+
+/**
+ * @brief A tuple which has the info related to the occured inotify event
+ *
+ * std::string - name[] in inotify_event struct
+ * uint32_t    - Mask describing event
+ */
+using BaseName = std::string;
+using EventMask = uint32_t;
+using EventInfo = std::tuple<BaseName, EventMask>;
+
+/** @class FD
+ *
+ *  @brief RAII wrapper for file descriptor.
+ */
+class FD
+{
+  private:
+    /** @brief File descriptor */
+    int fd = -1;
+
+  public:
+    FD(const FD&) = delete;
+    FD& operator=(const FD&) = delete;
+    FD(FD&&) = delete;
+    FD& operator=(FD&&) = delete;
+
+    /** @brief Saves File descriptor and uses it to do file operation
+     *
+     *  @param[in] fd - File descriptor
+     */
+    FD(int fd) : fd(fd) {}
```

Comment:
nit:  The cppcheck tool that runs in CI (but doesn't cause fails) will complain if single argument constructors aren't explicit.  See https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#c46-by-default-declare-single-argument-constructors-explicit